In [1]:
import numpy as np
import pandas as pd

### Question 1
What's the version of NumPy that you installed?

In [2]:
print(np.__version__)

1.19.5


### Fetching car dataset

```bash
https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
```

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv')

### Question 2
How many records are in the dataset?

In [5]:
df.shape

(11914, 16)

The dataset has 11,914 rows (instances) and 16 columns (features).

### Question 3
Who are the most popular car manufacturers (top-3) according to the dataset?

In [6]:
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [8]:
df.Make.value_counts()[0:3]

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

### Question 4
What's the number of unique Audi car models in the dataset?

In [11]:
df.loc[df.Make == 'Audi'].Model.nunique()

34

### Question 5
How many columns in the dataset have missing values?

In [12]:
df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

`Engine Fuel Type`, `Engine HP`, `Engine Cylinders`, `Number of Doors` and `Market Category` have missing values.

### Question 6
* Find the median value of "Engine Cylinders" column in the dataset.
* Next, calculate the most frequent value of the same "Engine Cylinders".
* Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
* Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

In [13]:
df['Engine Cylinders'].median()

6.0

In [14]:
df['Engine Cylinders'].mode()

0    4.0
Name: Engine Cylinders, dtype: float64

In [15]:
df['Engine Cylinders'].fillna(df['Engine Cylinders'].mode()[0], inplace=True)

In [16]:
df['Engine Cylinders'].median()

6.0

The median value did not change.

### Question 7
* Select all the `"Lotus"` cars from the dataset.
* Select only columns `"Engine HP", "Engine Cylinders"`.
* Now drop all duplicated rows using `drop_duplicates` method (you should get a dataframe with 9 rows).
* Get the underlying NumPy array. Let's call it `X`.
* Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
* Invert `XTX`.
* Create an array y with values `[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]`.
Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
What's the value of the first element of `w`?

In [18]:
X = df.loc[df.Make == 'Lotus'][["Engine HP", "Engine Cylinders"]].drop_duplicates().to_numpy()
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [19]:
XTX = (X.T).dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [20]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [21]:
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

In [22]:
w = XTX_inv.dot(X.T).dot(y)
w

array([  4.59494481, -63.56432501])